jax/_src/numpy/lax_numpy.py

```diff
@@ -1368,6 +1368,87 @@ def gradient(f, *varargs, axis: Optional[Union[int, Tuple[int, ...]]] = None,
 def isrealobj(x):
   return not iscomplexobj(x)
 
+@partial(jit, static_argnums=1)
+class _RankWarning(UserWarning):
+    """
+    Issued by `polyfit` when the Vandermonde matrix is rank deficient.
+    For more information, a way to suppress the warning, and an example of
+    `RankWarning` being issued, see `polyfit`.
+    """
+    pass
+
+@_wraps(np.polyfit)
+def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
+  _check_arraylike("polyfit", x, y, deg)
+  x = core.concrete_or_error(None, x, "The error arose in polyfit()")
+  y = core.concrete_or_error(None, y, "The error arose in polyfit()")
+  order = int(deg) + 1
+  # check arguments.
+  if deg < 0:
+      raise ValueError("expected deg >= 0")
+  if x.ndim != 1:
+      raise TypeError("expected 1D vector for x")
+  if x.size == 0:
+      raise TypeError("expected non-empty vector for x")
+  if y.ndim < 1 or y.ndim > 2:
+      raise TypeError("expected 1D or 2D array for y")
+  if x.shape[0] != y.shape[0]:
+      raise TypeError("expected x and y to have same length")
+
+  # set rcond
+  if rcond is None:
+      rcond = len(x)*finfo(x.dtype).eps
+
+  # set up least squares equation for powers of x
+  lhs = vander(x, order)
+  rhs = y
+
+  # apply weighting
+  if w is not None:
+      w = core.concrete_or_error(None, w, "The error arose in polyfit()") + 0.0
+      if w.ndim != 1:
+          raise TypeError("expected a 1-d array for weights")
+      if w.shape[0] != y.shape[0]:
+          raise TypeError("expected w and y to have the same length")
+      lhs *= w[:, newaxis]
+      if rhs.ndim == 2:
+          rhs *= w[:, newaxis]
+      else:
+          rhs *= w
+
+  # scale lhs to improve condition number and solve
+  scale = lax.sqrt((lhs*lhs).sum(axis=0))
+  lhs /= scale
+  c, resids, rank, s = lax.lstsq(lhs, rhs, rcond)
+  c = (c.T/scale).T  # broadcast scale coefficients
+
+  # warn on rank reduction, which indicates an ill conditioned matrix
+  if rank != order and not full:
```

Comment:
This statement will have to be removed for this function to be compatible with JIT (`rank` is derived from traced quantities, so we can't have logic based on its value). I'd consider removing this block and noting in the updated doc string that unlike numpy, the function will not warn if the input is poorly conditioned.